File: src/ros2_bridge/mappers/shape_msgs/solid_primitive.rs
//! Typed mapper for `shape_msgs/msg/SolidPrimitive`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn solid_primitive_to_bus(msg: ros_env::shape_msgs::msg::SolidPrimitive) -> crate::shape_msgs::msg::v1::SolidPrimitive {
    crate::shape_msgs::msg::v1::SolidPrimitive {
        r#type: msg.type_.into(),
        dimensions: crate::ros2_bridge::mappers::convert::f64_seq(msg.dimensions),
        polygon: Some(crate::ros2_bridge::mappers::geometry_msgs::polygon::polygon_to_bus(msg.polygon)),
    }
}

pub(crate) fn solid_primitive_to_ros(bus: crate::shape_msgs::msg::v1::SolidPrimitive) -> ros_env::shape_msgs::msg::SolidPrimitive {
    ros_env::shape_msgs::msg::SolidPrimitive {
        type_: bus.r#type as _,
        dimensions: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.dimensions),
        polygon: crate::ros2_bridge::mappers::geometry_msgs::polygon::polygon_to_ros(bus.polygon.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct ShapeMsgsSolidPrimitiveMapper;

impl TypedTopicMapper for ShapeMsgsSolidPrimitiveMapper {
    type Ros = ros_env::shape_msgs::msg::SolidPrimitive;
    type Bus = crate::shape_msgs::msg::v1::SolidPrimitive;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(solid_primitive_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(solid_primitive_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/shape_msgs/solid_primitive.py
"""Generated mapper for `shape_msgs/msg/SolidPrimitive`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.geometry_msgs.polygon import polygon_to_bus, polygon_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.shape_msgs.msg.v1 import SolidPrimitive as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def solid_primitive_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.type = msg.type
    bus.dimensions.extend(list(msg.dimensions))
    bus.polygon.CopyFrom(polygon_to_bus(msg.polygon))
    return bus


def solid_primitive_to_ros(bus):
    from shape_msgs.msg import SolidPrimitive as RosMsg

    out = RosMsg()
    out.type = bus.type
    out.dimensions = list(bus.dimensions)
    out.polygon = polygon_to_ros(bus.polygon)
    return out


class ShapeMsgsSolidPrimitiveMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from shape_msgs.msg import SolidPrimitive as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return solid_primitive_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return solid_primitive_to_ros(bus)
